File: backend/src/main/scala/domains/contest/objects/ContestProblemSummary.scala
package domains.contest.objects

import domains.problem.objects.{ProblemId, ProblemSlug, ProblemTitle}
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}

/** 比赛题目摘要，包含赛内位置和别名，用于详情页与榜单结果。 */
final case class ContestProblemSummary(
  id: ProblemId,
  slug: ProblemSlug,
  title: ProblemTitle,
  position: Int,
  alias: ContestProblemAlias
)

/** 提供比赛题目摘要 JSON codec。 */
object ContestProblemSummary:
  given Encoder[ContestProblemSummary] = deriveEncoder[ContestProblemSummary]
  given Decoder[ContestProblemSummary] = deriveDecoder[ContestProblemSummary]

File: frontend/src/objects/contest/ContestProblemSummary.ts
import type { ProblemId } from '@/objects/problem/ProblemId'
import type { ProblemSlug } from '@/objects/problem/ProblemSlug'
import type { ProblemTitle } from '@/objects/problem/ProblemTitle'
import type { ContestProblemAlias } from '@/objects/contest/ContestProblemAlias'

/** 比赛题目摘要；position 和 alias 决定比赛内展示顺序与题号。 */
export type ContestProblemSummary = {
  id: ProblemId
  slug: ProblemSlug
  title: ProblemTitle
  position: number
  alias: ContestProblemAlias
}
